### PR TITLE
feat: add whitebox module support for linux(#4654)

### PR DESCRIPTION
**Pull request by pollend**
there were a couple compile errors that I had to resolve but beyond that this seems to work mostly out of the box. the stl methods didn't compile correctly so I just reworked them to work with AzSTL. 

https://user-images.githubusercontent.com/854359/139184532-f122e4c8-db6f-4fbf-a1a5-4808ce5e731b.mp4

Only thing left would be to setup a build for a openmesh 3rd party dependency?

Signed-off-by: Michael Pollind <mpollind@gmail.com>

### CODE REVIEW

File: Gems/WhiteBox/Code/Source/Platform/Linux/platform_linux_tools.cmake
Comment (spham-amzn) [inline Gems/WhiteBox/Code/Source/Platform/Linux/platform_linux_tools.cmake:11]: The Linux package is now available. This file should also have been updated in the development branch to reflect the new package (for linux) and its new hash

```suggestion
    ly_associate_package(PACKAGE_NAME OpenMesh-8.1-rev3-linux TARGETS OpenMesh PACKAGE_HASH 805bd0b24911bb00c7f575b8c3f10d7ea16548a5014c40811894a9445f17a126)

```

File: Gems/WhiteBox/Code/Source/Platform/Linux/platform_linux_tools.cmake
Comment (spham-amzn) [reply Gems/WhiteBox/Code/Source/Platform/Linux/platform_linux_tools.cmake:13]: With the updated package mentioned above, this find_package is no longer needed

**Review by spham-amzn — APPROVED** [review]

File: Gems/WhiteBox/Code/Source/Core/WhiteBoxToolApi.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/Core/WhiteBoxToolApi.cpp:2981]: suggestion: Sorry I just noticed this but can we add the transform and reduce comments back please, I think they're helpful to know what each lambda is doi

File: Gems/WhiteBox/Code/Source/Viewport/WhiteBoxEdgeTranslationModifier.cpp
Comment (hultonha) [inline Gems/WhiteBox/Code/Source/Viewport/WhiteBoxEdgeTranslationModifier.cpp:71]: nit: It looks like there's trailing whitespace here (please could we remove) and I think we can leave the `cbegin` and `cend` calls

File: Gems/WhiteBox/Code/Source/Viewport/WhiteBoxEdgeTranslationModifier.cpp
Comment (hultonha) [reply Gems/WhiteBox/Code/Source/Viewport/WhiteBoxEdgeTranslationModifier.cpp:77]: nit: Can we please remove the trailing white space before merging

**Review by hultonha — APPROVED** [review]
I just noticed a few tiny things so fix-up, and now @spham-amzn's changes are in this should be good to merge very shortly!

File: Gems/WhiteBox/Code/Tests/WhiteBoxTestRailsAutomation.cpp
Comment (amznestebanpapp) [reply Gems/WhiteBox/Code/Tests/WhiteBoxTestRailsAutomation.cpp:161]: yeah, newer versions of MSVC are getting better at detecting unused, but clang detects way more... it also detects unused static functions and members

**Review by amznestebanpapp — APPROVED** [review]

**Review by lawsonamzn — APPROVED** [review]

**Review by hultonha — APPROVED** [review]

**Comment by hultonha** [comment]
I've just kicked off a build, once this goes green we should be good to merge (as it's Friday afternoon here I'll do this first thing on Monday! 🙂👍)

**Comment by hultonha** [comment]
https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-5075/2/

**Comment by spham-amzn** [comment]
> 
> 
> https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-5075/2/

The build failed but it is due to development still using clang-6. clang-12 has been enabled in stabilization, so we can either wait for a sync of development from stabilization, or @pollend can address `Wrange-loop-analysis` errors from the failure

**Comment by pollend** [comment]
> > https://jenkins.build.o3de.org/job/O3DE/view/change-requests/job/PR-5075/2/
> 
> The build failed but it is due to development still using clang-6. clang-12 has been enabled in stabilization, so we can either wait for a sync of development from stabilization, or @pollend can address `Wrange-loop-analysis` errors from the failure

pretty simple change I can do it tonight.

**Comment by hultonha** [comment]
Just kicked off another build. @spham-amzn can you confirm this is okay to merge this when it's done building? Thanks!

**Comment by hultonha** [comment]
Argh sorry about those additional issues, let me start another build 👍

**Comment by pollend** [comment]
> Argh sorry about those additional issues, let me start another build +1

do I just run this python script to get these errors? 

`python/python.sh -u scripts/build/ci_build.py --platform Linux --type profile_nounity`

**Comment by hultonha** [comment]
> 
> 
> > Argh sorry about those additional issues, let me start another build +1
> 
> do I just run this python script to get these errors?
> 
> `python/python.sh -u scripts/build/ci_build.py --platform Linux --type profile_nounity`

I don't think so, that's for the build and automate tests afaik, it should just be... `cmake --build build/linux --config profile -j 8` I think?

**Comment by amznestebanpapp** [comment]
> > Argh sorry about those additional issues, let me start another build +1
> 
> do I just run this python script to get these errors?
> 
> `python/python.sh -u scripts/build/ci_build.py --platform Linux --type profile_nounity`

Yes, you can run that locally to reproduce what Jenkins is running.
It should be the same as `cmake --build build/linux --config profile -j 8` configuring the project without unity.
Jenkins scripts should work locally for most cases (there are some cases where is known it doesnt work without having certain things locally, like the `installer` job in Windows requires WIX

**Comment by pollend** [comment]
> > > Argh sorry about those additional issues, let me start another build +1
> > 
> > 
> > do I just run this python script to get these errors?
> > `python/python.sh -u scripts/build/ci_build.py --platform Linux --type profile_nounity`
> 
> Yes, you can run that locally to reproduce what Jenkins is running. It should be the same as `cmake --build build/linux --config profile -j 8` configuring the project without unity. Jenkins scripts should work locally for most cases (there are some cases where is known it doesnt work without having certain things locally, like the `installer` job in Windows requires WIX

I tried a couple different versions of clang so hopefully this is the last run :P

I'll have this here for reference later:
```
cmake .. -G 'Ninja Multi-Config' -DCMAKE_C_COMPILER=clang-13 -DCMAKE_CXX_COMPILER=clang++-13 -DLY_UNITY_BUILD=FALSE -DLY_PARALLEL_LINK_JOBS=4
cmake .. -G 'Ninja Multi-Config' -DCMAKE_C_COMPILER=clang-8 -DCMAKE_CXX_COMPILER=clang++-8 -DLY_UNITY_BUILD=FALSE -DLY_PARALLEL_LINK_JOBS=4
```


found some other bits of code with unused variables
```
diff --git a/Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/TableView.cpp b/Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/TableView.cpp
index 725349cb5..2d5271fd0 100644
--- a/Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/TableView.cpp
+++ b/Code/Framework/AzQtComponents/AzQtComponents/Components/Widgets/TableView.cpp
@@ -213,12 +213,12 @@ namespace AzQtComponents
         // rectangle drawing.
         if (qTableView || qListView)
         {
-            int hHdr = 0;
+            // int hHdr = 0;
             int vHdr = 0;
 
             if (qTableView)
             {
-                hHdr = qTableView->horizontalHeader()->isVisible() ? qTableView->horizontalHeader()->height() : 0;
+                qTableView->horizontalHeader()->isVisible() ? qTableView->horizontalHeader()->height() : 0;
                 vHdr = qTableView->verticalHeader()->isVisible() ? qTableView->verticalHeader()->width() : 0;
             }
 
diff --git a/Code/Framework/AzToolsFramework/AzToolsFramework/UI/LegacyFramework/MainWindowSavedState.cpp b/Code/Framework/AzToolsFramework/AzToolsFramework/UI/LegacyFramework/MainWindowSavedState.cpp
index a9af59f3a..a73cbd09d 100644
--- a/Code/Framework/AzToolsFramework/AzToolsFramework/UI/LegacyFramework/MainWindowSavedState.cpp
+++ b/Code/Framework/AzToolsFramework/AzToolsFramework/UI/LegacyFramework/MainWindowSavedState.cpp
@@ -23,7 +23,7 @@ namespace AzToolsFramework
         // save 2k at a time :( need a better way to do this.
         AZStd::size_t pos = 0;
         AZStd::size_t remaining = m_windowState.size();
-        AZ::u8* charData = (AZ::u8*)windowState.begin();
+        // AZ::u8* charData = (AZ::u8*)windowState.begin();
 
         while (remaining > 0)
         {
@@ -31,7 +31,7 @@ namespace AzToolsFramework
             m_serializableWindowState.push_back();
             m_serializableWindowState.back().assign((AZ::u8*)windowState.begin() + pos, (AZ::u8*)windowState.begin() + pos + bytes_this_gulp);
             pos += bytes_this_gulp;
-            charData += bytes_this_gulp;
+            // charData += bytes_this_gulp;
             remaining -= bytes_this_gulp;
         }
     }
diff --git a/Code/Framework/AzToolsFramework/AzToolsFramework/UI/Slice/SlicePushWidget.cpp b/Code/Framework/AzToolsFramework/AzToolsFramework/UI/Slice/SlicePushWidget.cpp
index 76f91b438..b61530be0 100644
--- a/Code/Framework/AzToolsFramework/AzToolsFramework/UI/Slice/SlicePushWidget.cpp
+++ b/Code/Framework/AzToolsFramework/AzToolsFramework/UI/Slice/SlicePushWidget.cpp
@@ -1201,7 +1201,7 @@ namespace AzToolsFramework
                     .arg((item->parent() == nullptr) ? item->m_entity->GetName().c_str() : GetNodeDisplayName(*item->m_node).c_str()));
             }
 
-            SliceTargetTreeItem* parent = nullptr;
+            // SliceTargetTreeItem* parent = nullptr;
 
             AZStd::vector<SliceAssetPtr> validSliceAssets = GetValidTargetAssetsForField(*item);
 
@@ -1274,7 +1274,7 @@ namespace AzToolsFramework
                     selectButton->setChecked(true);
                 }
 
-                parent = sliceItem;
+                // parent = sliceItem;
                 ++level;
             }
         }
diff --git a/Code/Legacy/CryCommon/WinBase.cpp b/Code/Legacy/CryCommon/WinBase.cpp
index 9f7cdfd60..a3dd3a284 100644
--- a/Code/Legacy/CryCommon/WinBase.cpp
+++ b/Code/Legacy/CryCommon/WinBase.cpp
@@ -1282,27 +1282,27 @@ const bool GetFilenameNoCase
         }
     }
 
-    char* slash;
-    const char* dirname;
-    char* name;
-
-    if ((pAdjustedFilename) == (char*)-1)
-    {
-        return false;
-    }
-
-    slash = strrchr(pAdjustedFilename, '/');
-    if (slash)
-    {
-        dirname = pAdjustedFilename;
-        name = slash + 1;
-        *slash = 0;
-    }
-    else
-    {
-        dirname = ".";
-        name = pAdjustedFilename;
-    }
+    // char* slash;
+    // const char* dirname;
+    // char* name;
+
+    // if ((pAdjustedFilename) == (char*)-1)
+    // {
+    //     return false;
+    // }
+
+    // slash = strrchr(pAdjustedFilename, '/');
+    // if (slash)
+    // {
+    //     dirname = pAdjustedFilename;
+    //     name = slash + 1;
+    //     *slash = 0;
+    // }
+    // else
+    // {
+    //     dirname = ".";
+    //     name = pAdjustedFilename;
+    // }
 
 #if !defined(LINUX) && !defined(APPLE) && !defined(DEFINE_SKIP_WILDCARD_CHECK)      // fix the parent path anyhow.
     // Check for wildcards. We'll always return true if the specified filename is
@@ -1318,10 +1318,10 @@ const bool GetFilenameNoCase
 #endif
 
     // Scan for the file.
-    if (slash)
-    {
-        *slash = '/';
-    }
+    // if (slash)
+    // {
+    //     *slash = '/';
+    // }

```

**Comment by hultonha** [comment]
Kicked off another build 😅 We'll get there I'm sure, thanks for persevering @pollend!

**Comment by hultonha** [comment]
Apologies it looks like it failed again - https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/5/console. Have you already fixed these warnings now @pollend? I can kick off another PR shortly

**Comment by pollend** [comment]
> Apologies it looks like it failed again - https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/5/console. Have you already fixed these warnings now @pollend? I can kick off another PR shortly

looks like I missed building the test cases.

**Comment by pollend** [comment]
> Apologies it looks like it failed again - https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/5/console. Have you already fixed these warnings now @pollend? I can kick off another PR shortly

I guess there are some failing tests. I can try resolve them tonight.

**Comment by hultonha** [comment]
I'm afraid so @pollend, I see in the log [here](https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/6/consoleFull#9928285060f5c32df5baa2a40af90ad8)

```
18:03:04  	 92 - Gem::WhiteBox.Editor.Tests.main::TEST_RUN (Failed)
18:03:04  	 93 - Gem::WhiteBox.Editor.Physics.Tests.main::TEST_RUN (Failed)
18:03:04  Errors while running CTest
```

and this error

```
Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
17:59:48  Returning code: 102
```

And the same here...

```
Loading: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
17:59:48  Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
17:59:48  Returning code: 102
```

Not sure why this is getting flagged up... @spham-amzn might have an idea?

**Comment by pollend** [comment]
> I'm afraid so @pollend, I see in the log [here](https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/6/consoleFull#9928285060f5c32df5baa2a40af90ad8)
> 
> ```
> 18:03:04  	 92 - Gem::WhiteBox.Editor.Tests.main::TEST_RUN (Failed)
> 18:03:04  	 93 - Gem::WhiteBox.Editor.Physics.Tests.main::TEST_RUN (Failed)
> 18:03:04  Errors while running CTest
> ```
> 
> and this error
> 
> ```
> Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> 17:59:48  Returning code: 102
> ```
> 
> And the same here...
> 
> ```
> Loading: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Returning code: 102
> ```
> 
> Not sure why this is getting flagged up... @spham-amzn might have an idea?

whats  different about the other gems that also have tests.

**Comment by spham-amzn** [comment]
> > I'm afraid so @pollend, I see in the log [here](https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/6/consoleFull#9928285060f5c32df5baa2a40af90ad8)
> > ```
> > 18:03:04  	 92 - Gem::WhiteBox.Editor.Tests.main::TEST_RUN (Failed)
> > 18:03:04  	 93 - Gem::WhiteBox.Editor.Physics.Tests.main::TEST_RUN (Failed)
> > 18:03:04  Errors while running CTest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and this error
> > ```
> > Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> > 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> > 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> > 17:59:48  Returning code: 102
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And the same here...
> > ```
> > Loading: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> > 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Returning code: 102
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure why this is getting flagged up... @spham-amzn might have an idea?
> 
> whats different about the other gems that also have tests.

I'll look into it

**Comment by pollend** [comment]
> I'm afraid so @pollend, I see in the log [here](https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/6/consoleFull#9928285060f5c32df5baa2a40af90ad8)
> 
> ```
> 18:03:04  	 92 - Gem::WhiteBox.Editor.Tests.main::TEST_RUN (Failed)
> 18:03:04  	 93 - Gem::WhiteBox.Editor.Physics.Tests.main::TEST_RUN (Failed)
> 18:03:04  Errors while running CTest
> ```
> 
> and this error
> 
> ```
> Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> 17:59:48  Returning code: 102
> ```
> 
> And the same here...
> 
> ```
> Loading: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> 17:59:48  Returning code: 102
> ```
> 
> Not sure why this is getting flagged up... @spham-amzn might have an idea?

I tried to run it on my machine and it seems to work fine for both tests. I'm running a newer version of clang and cmake. The current version of ubuntu does not even let me install clang-6. was openmesh built on newer packages? everything else should be linking to similar packages.

**Review by cgalvan — APPROVED** [review]

**Comment by hultonha** [comment]
It does look like that worked! 🥳

I'll get this merged first thing on Monday! Thank you so much for all your time, effort (and patience!) @pollend, glad we got there in the end 🙂

**Comment by hultonha** [comment]
Merged! 🥳

**Comment by spham-amzn** [comment]
> 
> 
> > I'm afraid so @pollend, I see in the log [here](https://jenkins.build.o3de.org/job/O3DE/job/PR-5075/6/consoleFull#9928285060f5c32df5baa2a40af90ad8)
> > ```
> > 18:03:04  	 92 - Gem::WhiteBox.Editor.Tests.main::TEST_RUN (Failed)
> > 18:03:04  	 93 - Gem::WhiteBox.Editor.Physics.Tests.main::TEST_RUN (Failed)
> > 18:03:04  Errors while running CTest
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and this error
> > ```
> > Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> > 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> > 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Tests.so
> > 17:59:48  Returning code: 102
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And the same here...
> > ```
> > Loading: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Fail to open shared library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Error from dlopen(): /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so: undefined symbol: _ZNSt7__cxx1119basic_ostringstreamIcSt11char_traitsIcESaIcEEC1Ev
> > 17:59:48  FAILED to load library: /data/workspace/o3de/build/linux/bin/profile/libWhiteBox.Editor.Physics.Tests.so
> > 17:59:48  Returning code: 102
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Not sure why this is getting flagged up... @spham-amzn might have an idea?
> 
> I tried to run it on my machine and it seems to work fine for both tests. I'm running a newer version of clang and cmake. The current version of ubuntu does not even let me install clang-6. was openmesh built on newer packages? everything else should be linking to similar packages.
> 
> looks like the latest develop is using a newer version of clang let me rebase this.

The jenkins jobs that were running the development branch was only recently updated to clang-12, so that would explain it

**Pull request merged**
